### PR TITLE
adding jshint rake task and dev dependency

### PR DESCRIPTION
**Pull request by lzap**
SSIA

After review I can go ahead and fix the javascript codebase. There are
plenty of errors tho.

### CODE REVIEW

**Comment by lzap** [comment]
```
[lzap@lzapx foreman]$ be rake jshint
file: app/assets/javascripts/about.js
  line 3, col 14, 'item' is already defined.
  line 11, col 28, Expected '===' and instead saw '=='.
  line 12, col 47, Missing semicolon.
  line 14, col 49, Missing semicolon.
  line 18, col 7, Missing semicolon.
  line 19, col 5, Missing semicolon.
file: app/assets/javascripts/application.js
  line 16, col 23, 'onContentLoad' was used before it was defined.
  line 20, col 54, Expected '===' and instead saw '=='.
  line 21, col 46, Expected '===' and instead saw '=='.
  line 50, col 5, Missing semicolon.
  line 78, col 30, 'mark_params_override' was used before it was defined.
  line 84, col 28, Expected '===' and instead saw '=='.
  line 86, col 74, Missing semicolon.
  line 89, col 7, Missing semicolon.
  line 90, col 5, Missing semicolon.
  line 98, col 54, Expected '===' and instead saw '=='.
  line 98, col 108, Expected '===' and instead saw '=='.
  line 108, col 61, Expected '{' and instead saw '$'.
  line 135, col 34, Unexpected use of '++'.
  line 136, col 34, Unexpected use of '++'.
  line 166, col 28, Expected '===' and instead saw '=='.
  line 237, col 91, Missing semicolon.
  line 246, col 12, Expected '===' and instead saw '=='.
  line 252, col 12, Expected '===' and instead saw '=='.
  line 258, col 12, Expected '===' and instead saw '=='.
  line 264, col 42, Expected '===' and instead saw '=='.
  line 264, col 85, Expected '===' and instead saw '=='.
  line 274, col 22, Expected '===' and instead saw '=='.
  line 283, col 17, Expected '===' and instead saw '=='.
  line 283, col 42, Expected '===' and instead saw '=='.
  line 292, col 36, Unexpected use of '++'.
  line 295, col 27, Expected '===' and instead saw '=='.
  line 299, col 10, Don't make functions within a loop.
  line 299, col 11, Missing semicolon.
  line 301, col 24, Expected '!==' and instead saw '!='.
  line 301, col 33, Expected '{' and instead saw 'attrs'.
  line 327, col 37, Missing semicolon.
  line 331, col 38, Missing semicolon.
  line 332, col 14, Expected '===' and instead saw '=='.
  line 332, col 21, Expected '{' and instead saw 'return'.
  line 345, col 5, Missing semicolon.
  line 357, col 18, Expected '===' and instead saw '=='.
  line 366, col 2, Missing semicolon.
  line 370, col 2, Missing semicolon.
  line 372, col 29, 'spinner_placeholder' was used before it was defined.
  line 373, col 12, Expected '===' and instead saw '=='.
  line 373, col 26, Expected '{' and instead saw 'text'.
  line 374, col 63, Missing semicolon.
file: app/assets/javascripts/charts.js
  line 41, col 2, Missing semicolon.
  line 102, col 2, Missing semicolon.
  line 132, col 4, Missing semicolon.
  line 142, col 2, Missing semicolon.
  line 144, col 30, 'chart_legend_options' was used before it was defined.
  line 145, col 33, Expected '===' and instead saw '=='.
  line 145, col 39, Expected '{' and instead saw 'return'.
  line 156, col 8, Missing semicolon.
  line 164, col 19, 'flot_zoom' was used before it was defined.
  line 167, col 5, Expected '{' and instead saw 'ranges'.
  line 169, col 5, Expected '{' and instead saw 'ranges'.
  line 177, col 63, Expected '===' and instead saw '=='.
  line 198, col 26, 'bind_hover_event' was used before it was defined.
  line 201, col 26, Expected '!==' and instead saw '!='.
  line 213, col 25, 'search_on_click' was used before it was defined.
  line 215, col 12, Expected '===' and instead saw '=='.
  line 215, col 26, Expected '{' and instead saw 'return'.
  line 216, col 40, Expected '!==' and instead saw '!='.
  line 218, col 95, Missing semicolon.
  line 220, col 32, Expected '!==' and instead saw '!='.
  line 226, col 16, 'val1' is already defined.
  line 227, col 29, Expected '!==' and instead saw '!='.
  line 227, col 36, Expected '{' and instead saw 'val1'.
  line 229, col 35, 'val1' used out of scope.
  line 236, col 24, Expected '===' and instead saw '=='.
  line 244, col 40, Missing semicolon.
  line 256, col 29, Expected '===' and instead saw '=='.
  line 277, col 41, Unexpected use of '++'.
  line 299, col 41, Unexpected use of '++'.
file: app/assets/javascripts/compute_resource.js
  line 6, col 18, Expected '===' and instead saw '=='.
  line 18, col 15, Expected '===' and instead saw '=='.
  line 71, col 7, Missing semicolon.
  line 74, col 31, 'add_network_interface' was used before it was defined.
  line 81, col 20, 'add_volume' was used before it was defined.
  line 90, col 25, 'disable_element' was used before it was defined.
  line 122, col 7, Missing semicolon.
  line 147, col 34, 'disable_libvirt_dropdown' was used before it was defined.
  line 152, col 33, 'enable_libvirt_dropdown' was used before it was defined.
  line 154, col 45, Missing semicolon.
file: app/assets/javascripts/config_template.js
  line 1, col 12, Missing semicolon.
  line 11, col 67, Unnecessary semicolon.
  line 19, col 7, Missing semicolon.
  line 23, col 35, Expected '===' and instead saw '=='.
  line 23, col 42, Expected '{' and instead saw 'return'.
  line 27, col 93, Missing semicolon.
  line 31, col 38, Expected a conditional expression and instead saw an assignment.
  line 31, col 41, Unexpected use of '++'.
  line 35, col 37, Expected '===' and instead saw '=='.
  line 39, col 10, Don't make functions within a loop.
  line 50, col 5, Missing semicolon.
  line 61, col 5, Missing semicolon.
  line 70, col 23, 'create_editor' was used before it was defined.
  line 87, col 48, Expected '{' and instead saw 'return'.
  line 90, col 37, Missing semicolon.
  line 98, col 23, 'set_edit_mode' was used before it was defined.
  line 111, col 23, 'set_diff_mode' was used before it was defined.
  line 118, col 20, Expected '===' and instead saw '=='.
  line 119, col 5, Expected '{' and instead saw 'patch'.
  line 119, col 25, Missing semicolon.
  line 125, col 22, 'IE_diff_mode' was used before it was defined.
  line 132, col 113, Missing semicolon.
  line 133, col 16, Expected '{' and instead saw 'return'.
  line 150, col 100, Missing semicolon.
  line 152, col 5, Missing semicolon.
file: app/assets/javascripts/host_checkbox.js
  line 7, col 53, Missing radix parameter.
  line 9, col 5, Expected '{' and instead saw 'addHostId'.
  line 11, col 5, Expected '{' and instead saw 'rmHostId'.
  line 18, col 19, 'addHostId' was used before it was defined.
  line 19, col 50, Expected '===' and instead saw '=='.
  line 20, col 5, Expected '{' and instead saw '$'.
  line 20, col 36, Missing semicolon.
  line 23, col 18, 'rmHostId' was used before it was defined.
  line 26, col 5, Expected '{' and instead saw '$'.
  line 26, col 42, Missing semicolon.
  line 29, col 24, 'readFromCookie' was used before it was defined.
  line 31, col 35, Expected a conditional expression and instead saw an assignment.
  line 32, col 7, Expected '{' and instead saw 'return'.
  line 34, col 7, Expected '{' and instead saw 'return'.
  line 42, col 24, 'toggle_actions' was used before it was defined.
  line 44, col 37, Expected '===' and instead saw '=='.
  line 55, col 55, Unexpected use of '++'.
  line 58, col 7, Expected '{' and instead saw 'boxes'.
  line 65, col 34, 'removeForemanHostsCookie' was used before it was defined.
  line 92, col 5, Expected '{' and instead saw 'cleanHostsSelection'.
  line 98, col 18, Expected '!==' and instead saw '!='.
  line 99, col 5, Expected '{' and instead saw 'b'.
  line 101, col 5, Expected '{' and instead saw 'b'.
  line 107, col 39, Expected '===' and instead saw '=='.
  line 119, col 13, Expected '{' and instead saw 'b'.
  line 121, col 13, Expected '{' and instead saw 'b'.
  line 137, col 24, 'update_counter' was used before it was defined.
  line 145, col 5, Expected '{' and instead saw 'title'.
  line 147, col 5, Expected '{' and instead saw 'title'.
file: app/assets/javascripts/host_edit.js
  line 3, col 13, Expected '===' and instead saw '=='.
  line 33, col 69, Expected '{' and instead saw '$'.
  line 36, col 7, Missing semicolon.
  line 40, col 29, 'update_capabilities' was used before it was defined.
  line 98, col 22, 'clear_errors' was used before it was defined.
  line 105, col 26, 'animate_progress' was used before it was defined.
  line 106, col 20, Expected '===' and instead saw '=='.
  line 106, col 29, Expected '{' and instead saw 'return'.
  line 112, col 7, Missing semicolon.
  line 116, col 25, 'update_progress' was used before it was defined.
  line 118, col 22, Expected '===' and instead saw '=='.
  line 118, col 43, Expected '===' and instead saw '=='.
  line 118, col 52, Expected '{' and instead saw 'return'.
  line 180, col 61, Expected '{' and instead saw '$'.
  line 185, col 38, 'load_puppet_class_parameters' was used before it was defined.
  line 187, col 37, Missing semicolon.
  line 188, col 63, Expected '{' and instead saw 'return'.
  line 189, col 66, Expected '{' and instead saw 'return'.
  line 192, col 38, Missing semicolon.
  line 194, col 11, Expected '===' and instead saw '=='.
  line 194, col 25, Expected '{' and instead saw 'return'.
  line 206, col 45, Expected '{' and instead saw '$'.
  line 235, col 21, 'update_form' was used before it was defined.
  line 239, col 21, Expected '{' and instead saw 'data'.
  line 255, col 5, Missing semicolon.
  line 260, col 17, Expected '===' and instead saw '=='.
  line 260, col 47, Expected '===' and instead saw '=='.
  line 260, col 53, Expected '{' and instead saw 'return'.
  line 264, col 24, Expected '!==' and instead saw '!='.
  line 264, col 70, Expected '!==' and instead saw '!='.
  line 267, col 7, Expected '{' and instead saw 'return'.
  line 280, col 5, Missing semicolon.
  line 283, col 25, 'subnet_contains' was used before it was defined.
  line 286, col 38, Missing radix parameter.
  line 287, col 18, Unexpected use of '>>'.
  line 287, col 41, Unexpected use of '>>'.
  line 287, col 27, Expected '===' and instead saw '=='.
  line 290, col 17, '_to_int' was used before it was defined.
  line 293, col 21, Unexpected use of '++'.
  line 294, col 51, Missing radix parameter.
  line 312, col 5, Missing semicolon.
  line 327, col 5, Missing semicolon.
  line 346, col 35, 'update_provisioning_image' was used before it was defined.
  line 350, col 18, Expected '===' and instead saw '=='.
  line 350, col 47, Expected '===' and instead saw '=='.
  line 350, col 66, Expected '===' and instead saw '=='.
  line 350, col 83, Expected '===' and instead saw '=='.
  line 350, col 91, Expected '{' and instead saw 'return'.
  line 363, col 11, Expected '{' and instead saw 'image_options'.
  line 365, col 7, Missing semicolon.
  line 365, col 7, Too many errors. (59% scanned).
  fatal error
file: app/assets/javascripts/i18n.js
  line 7, col 25, Expected '!==' and instead saw '!='.
file: app/assets/javascripts/lookup_keys.js
  line 10, col 3, Missing semicolon.
  line 12, col 26, 'select_first_tab' was used before it was defined.
  line 21, col 8, Missing semicolon.
  line 24, col 28, 'show_delete_button' was used before it was defined.
  line 34, col 21, 'remove_node' was used before it was defined.
  line 50, col 17, Expected '===' and instead saw '=='.
  line 50, col 67, Unnecessary semicolon.
  line 54, col 108, Bad or unnecessary escaping.
  line 54, col 116, Bad or unnecessary escaping.
  line 64, col 48, Unexpected use of '++'.
  line 82, col 15, Expected '===' and instead saw '=='.
  line 84, col 180, Missing semicolon.
  line 118, col 32, 'undo_remove_child_node' was used before it was defined.
  line 135, col 51, Expected '===' and instead saw '=='.
  line 160, col 20, Expected '===' and instead saw '=='.
  line 183, col 55, Expected '===' and instead saw '=='.
file: app/assets/javascripts/nfs_visibility.js
  line 3, col 51, Missing semicolon.
  line 6, col 50, Missing semicolon.
  line 7, col 5, Missing semicolon.
  line 12, col 23, Expected '===' and instead saw '=='.
  line 13, col 12, Unexpected use of '++'.
  line 16, col 15, Missing semicolon.
  line 18, col 28, 'toggle_nfs_section' was used before it was defined.
  line 26, col 27, 'check_nfs_section' was used before it was defined.
  line 28, col 37, Expected '===' and instead saw '=='.
  line 31, col 46, Missing semicolon.
file: app/assets/javascripts/noVNC.js
  line 1, col 1, Use the function form of "use strict".
  line 11, col 7, '$D' is not defined.
  line 12, col 8, '$D' is not defined.
  line 13, col 9, '$D' is not defined.
  line 32, col 3, '$D' is not defined.
  line 33, col 3, '$D' is not defined.
  line 39, col 13, 'RFB' is not defined.
  line 39, col 28, '$D' is not defined.
file: app/assets/javascripts/spice.js
  line 26, col 21, 'spice_error' was used before it was defined.
  line 32, col 23, 'spice_success' was used before it was defined.
  line 33, col 114, Missing semicolon.
  line 38, col 30, Expected '===' and instead saw '=='.
file: app/assets/javascripts/taxonomy.js
  line 4, col 5, Missing semicolon.
  line 5, col 3, Missing semicolon.
  line 7, col 24, 'ignore_checked' was used before it was defined.
file: app/assets/javascripts/topbar.js
  line 17, col 3, Missing semicolon.
  line 19, col 19, 'is_mobile' was used before it was defined.
  line 21, col 1965, Expected '{' and instead saw 'check'.
  line 25, col 18, 'onScroll' was used before it was defined.
file: app/assets/javascripts/trends.js
  line 2, col 55, Missing semicolon.
  line 6, col 27, 'trendTypeSelected' was used before it was defined.
  line 7, col 32, Expected '===' and instead saw '=='.
JSHint check failed
[lzap@lzapx foreman]$ echo $?
1
```

The return code can be used in CI.

**Comment by lzap** [comment]
@abenari do you see something on the list above that should be rather turned off in the jslint than fixed in our codebase?

File: lib/tasks/jshint.rake [hunk 1]
```diff
@@ -0,0 +1,35 @@
+if Rails.env.development?
```
Comment (domcleal) [inline lib/tasks/jshint.rake]: test too?

**Comment by domcleal** [comment]
Can we add this to the test task so CI runs it, similar to `lib/tasks/test_lib_dir.rake`?

**Comment by lzap** [comment]
Ok I managed to configure jshintrb to skip some obvious things that does not need to be done "right", but then I started fixing the rest and realized this is not worth the effort. Jshintrb is poor project, does not differentiate between warnings and errors and upstream author is not responsive.

I am closing this pull request, some time wasted.

```
$ be rake test:jshint
file: app/assets/javascripts/about.js
  line 3, col 14, 'item' is already defined.
file: app/assets/javascripts/application.js
  line 20, col 54, Use '===' to compare with '0'.
  line 21, col 46, Use '===' to compare with '0'.
  line 274, col 22, Use '===' to compare with ''.
  line 332, col 14, Use '===' to compare with ''.
  line 373, col 12, Use '===' to compare with 'undefined'.
file: app/assets/javascripts/charts.js
  line 177, col 63, Use '===' to compare with '0'.
  line 215, col 12, Use '===' to compare with 'undefined'.
  line 226, col 16, 'val1' is already defined.
  line 229, col 35, 'val1' used out of scope.
  line 236, col 24, Use '===' to compare with '0'.
  line 256, col 29, Use '===' to compare with '0'.
file: app/assets/javascripts/compute_resource.js
  line 18, col 15, Use '===' to compare with ''.
file: app/assets/javascripts/config_template.js
  line 11, col 67, Unnecessary semicolon.
  line 23, col 35, Use '===' to compare with ''.
  line 31, col 38, Expected a conditional expression and instead saw an assignment.
  line 118, col 20, Use '===' to compare with '0'.
file: app/assets/javascripts/host_checkbox.js
  line 7, col 53, Missing radix parameter.
  line 31, col 35, Expected a conditional expression and instead saw an assignment.
  line 44, col 37, Use '===' to compare with '0'.
  line 107, col 39, Use '===' to compare with '0'.
file: app/assets/javascripts/host_edit.js
  line 3, col 13, Use '===' to compare with ''.
  line 106, col 20, Use '===' to compare with 'true'.
  line 118, col 22, Use '===' to compare with '0'.
  line 118, col 43, Use '===' to compare with 'true'.
  line 194, col 11, Use '===' to compare with 'undefined'.
  line 260, col 17, Use '===' to compare with ''.
  line 260, col 47, Use '===' to compare with '0'.
  line 264, col 24, Use '!==' to compare with '0'.
  line 286, col 38, Missing radix parameter.
  line 287, col 18, Unexpected use of '>>'.
  line 287, col 41, Unexpected use of '>>'.
  line 294, col 51, Missing radix parameter.
  line 350, col 18, Use '===' to compare with 'undefined'.
  line 350, col 47, Use '===' to compare with ''.
  line 350, col 66, Use '===' to compare with ''.
  line 350, col 83, Use '===' to compare with ''.
  line 399, col 25, Use '===' to compare with ''.
  line 455, col 9, Use '===' to compare with 'undefined'.
  line 517, col 17, Use '===' to compare with ''.
  line 556, col 17, Use '===' to compare with ''.
  line 566, col 24, Use '!==' to compare with '0'.
file: app/assets/javascripts/lookup_keys.js
  line 50, col 17, Use '===' to compare with 'undefined'.
  line 50, col 67, Unnecessary semicolon.
  line 54, col 108, Bad or unnecessary escaping.
  line 54, col 116, Bad or unnecessary escaping.
  line 135, col 51, Use '===' to compare with '0'.
  line 160, col 20, Use '===' to compare with ''.
  line 183, col 55, Use '===' to compare with ''.
file: app/assets/javascripts/nfs_visibility.js
  line 12, col 23, Use '===' to compare with 'true'.
  line 28, col 37, Use '===' to compare with '0'.
file: app/assets/javascripts/noVNC.js
  line 1, col 1, Use the function form of "use strict".
  line 11, col 7, '$D' is not defined.
  line 12, col 8, '$D' is not defined.
  line 13, col 9, '$D' is not defined.
  line 32, col 3, '$D' is not defined.
  line 33, col 3, '$D' is not defined.
  line 39, col 13, 'RFB' is not defined.
  line 39, col 28, '$D' is not defined.
file: app/assets/javascripts/spice.js
  line 38, col 30, Use '===' to compare with '0'.
JSHint check failed
```

**Pull request closed without merging**